Userland/Libraries/LibGUI/GML/AST.h

```diff
@@ -0,0 +1,319 @@
+// FIXME: Use a specialized value type for all the possible GML property values. Right now that's all possible JSON values (?)
+class JsonValueNode : public ValueNode
+    , public JsonValue {
+
+public:
+    JsonValueNode(JsonValue const& value)
+        : JsonValue(value)
+    {
+    }
+
+    virtual void format(StringBuilder& builder, size_t indentation, bool is_inline) const override
+    {
+        if (!is_inline)
+            indent(builder, indentation);
+        if (is_array()) {
+            // custom array serialization as AK's doesn't pretty-print
+            // objects and arrays (we only care about arrays (for now))
+            builder.append("[");
+            auto first = true;
+            as_array().for_each([&](auto& value) {
+                if (!first)
+                    builder.append(", ");
+                first = false;
+                value.serialize(builder);
+            });
+            builder.append("]");
+        } else {
+            serialize(builder);
+        }
+        if (!is_inline)
+            builder.append("\n");
+    }
+};
+
+// GML class declaration, starting with '@'
+class Object : public ValueNode {
+public:
+    Object() = default;
+    Object(String name, NonnullRefPtrVector<Node> children)
+        : m_children(move(children))
+        , m_name(move(name))
+    {
+    }
+
+    virtual ~Object() override = default;
+
+    StringView name() const { return m_name; }
+    void set_name(String name) { m_name = move(name); }
+    NonnullRefPtrVector<Node> children() const { return m_children; }
+
+    ErrorOr<void> add_child(NonnullRefPtr<Node> child)
+    {
+        return m_children.try_append(move(child));
+    }
+
+    // Does not return key-value pair `layout: ...`!
+    template<typename Callback>
+    void for_each_property(Callback callback)
+    {
+        for (auto const& child : m_children) {
+            if (is<KeyValuePair>(child)) {
+                auto const& property = static_cast<KeyValuePair const&>(child);
+                if (property.key() != "layout" && is<JsonValueNode>(property.value().ptr()))
+                    callback(property.key(), AK::static_ptr_cast<JsonValueNode>(property.value()));
+            }
+        }
+    }
+
+    template<typename Callback>
+    void for_each_child_object(Callback callback)
+    {
+        for (NonnullRefPtr<Node> child : m_children) {
+            // doesn't capture layout as intended, as that's behind a kv-pair
+            if (is<Object>(child.ptr())) {
+                auto object = static_ptr_cast<Object>(child);
+                callback(object);
+            }
+        }
+    }
+
+    // Uses IterationDecision to allow the callback to interrupt the iteration, like a for-loop break.
+    template<typename Callback>
+    void for_each_child_object_interruptible(Callback callback)
+    {
+        for (NonnullRefPtr<Node> child : m_children) {
+            // doesn't capture layout as intended, as that's behind a kv-pair
```

Comment:
The way this code special cases layout objects feels a bit fragile. the fact that they are the only Non-Widget items registered for GML loadability (and communication with Inspector?) is kind of incidental at the moment, rather than a conscious design choice. The previous code relied on the fact that layouts would always be a key: value pair on a widget, and only recursed through `children` arrays, assuming that children would always be widgets...  Hrm. Seems the special casing logic has only moved and become more obvious rather than being a new problem. I don't think I'm recommending any action here, feels like a future problem.

Reply:
I do agree, however I don't see a solution right now. The AST needs to be a bit more general than even some of these GML semantics in order for e.g. comments to work properly.